Add FIRST survey images

rlwastro opened the pull request:
This commit adds the images from the VLA FIRST survey.  I modeled the code after the MAGPIS survey (which has a very similar interface on the same web server).

pep8speaks commented:
Hello @rlwastro! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:









* In the file [`astroquery/first/tests/test_first_remote.py`](https://github.com/astropy/astroquery/blob/88082dacd9f42357dc87d67674d7759a0251573f/astroquery/first/tests/test_first_remote.py):

> [Line 10:1](https://github.com/astropy/astroquery/blob/88082dacd9f42357dc87d67674d7759a0251573f/astroquery/first/tests/test_first_remote.py#L10): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1

keflavich commented:
Thanks.  Some of the object names are coincidentally amusing (`FirstClass`), but at a glance this looks all good (assuming travis passes).

One thing I'm slightly concerned about is the test file - @bsipocz, is a ~350 kB file OK to add, or is that likely to add substantial overhead, perhaps too much if extrapolated to many more services?  We could probably host that file at astropy.data if needed.

rlwastro commented:
Actually that file is a copy of a file that is already in the magpis/tests directory, so I'd be happy to omit it.  Or maybe it could be replaced with something small?  It is just a FITS file to use in the local testing.

keflavich commented:
oh, if it's the same file that's in another directory, just use that one.  We can even include a symlink  between the two

rlwastro commented:
The name of the FIRST survey was not so coincidentally chosen to facilitate amusement. :wink:

rlwastro commented:
I deleted the image.fits file and changed it to a link to the version in the magpis/tests directory. (Hope I did that correctly.)

rlwastro commented:
Looking at the the Travis CI error message for the failed Windows test, it looks like the image.fits link might not work correctly on Windows.  Is there a way to make a Windows-friendly link?  If not, I'll create a much smaller file to use for the test.

keflavich commented:
That's a great question.  @pllim, you're our local windows expert - can we use symlinks that will work with windows for CI?

@rlwastro , if not, just use the path to the MAGPIS file.  You can just set the DATA_PATH (https://github.com/astropy/astroquery/pull/1733/files#diff-c4d54f4dd2e496ccd9cfb6746d56227eR45) to point to the file in the MAGPIS directory.  I thought symlinking was easier, but maybe not.

DaftPict commented:
if I can interject - I've never been able to get symlinks on windows to work unless using a privileged account. There's supposed to be a policy change to allow this but ITSD has never been able to make it work for me..

pllim commented:
AFAIK, Windows doesn't like symlink.

rlwastro commented:
@keflavich I don't know how to trigger the Travis CI test again, but I'm guessing it will succeed this time.

pllim commented:
p.s. Interesting timing considering that WFIRST was just renamed... 😄

keflavich commented:
Could you rebase and cut out the commit where you added the big file originally?  Otherwise it gets included in the git history.

Anyway, this is probably good to merge with a few tiny modifications, we're just trying to keep the history and workspace clean - we probably want to put this and MAGPIS under a common parent directory, but that refactor has been on our to do list for a while.

rlwastro commented:
I rebased and believe I removed the commit that had the big file.  Let me know if I did this correctly.

I tried doing a force push as @pllim suggested to trigger the Travis CI testing, but I didn't get that to work.  Probably I'm not doing quite the right thing.  So I believe the final version has still not run through testing.

pllim commented:
To get rid of the merge commits and possible still needed to really get rid of the big file in history:

```
git rebase -i HEAD~N
```

where N is the number of commits you want to mess with. Then, follow instructions that pop up in the editor you registered with Git on your machined.

rlwastro commented:
The rebasing is done (with some help from @pllim ) so I think the work is complete.  And the Travis CI test ran this time and was successful, although it does not appear in the messages for some reason: [https://travis-ci.org/github/astropy/astroquery/builds/695174420](https://travis-ci.org/github/astropy/astroquery/builds/695174420).

pllim commented:
Yeah, I notice that Travis CI not reporting back on `astropy` repo as well.

bsipocz commented on astroquery/first/tests/test_first.py:
```diff
@@ -0,0 +1,63 @@
+        return c
+    try:
+        mp = request.getfixturevalue("monkeypatch")
+    except AttributeError:  # pytest < 3
```
we don't support python2.7 any more, so no need for the try/except

rlwastro replied:
Thanks, I made these changes (I inherited the Python2 compatibility code from the magpis module).

bsipocz commented on astroquery/first/tests/test_first.py:
```diff
@@ -0,0 +1,63 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+from __future__ import print_function
```
we live in the python3 future, no need for this :)

rlwastro replied:
Made these changes too

keflavich commented:
This is an interesting question, and @rlwastro, I'd value your input.  The logical places to live could be:
`vla/first` - then, though, we have the awkwardness that VLASS lives under CADC and NRAO has an archive, but the VLA doesn't
`nrao/first` - this probably makes more sense?

Or, should this live in `image_cutout_services/first` with MAGPIS?

See https://github.com/astropy/astroquery/pull/1451 for the proposed restructure.

bsipocz commented:
I could even live with a `radio/first` as a first approach, as we talked about having pseudo namespaces, too (we have that already for solar system, except here it could act as the fist location for the code).

bsipocz commented:
I also like your suggestion @keflavich  for `image_cutouts`, but maybe that's also a pseudo one.

rlwastro commented:
I can see the argument for making the namespace structure more hierarchical.  (To quote Tim Peters: Namespaces are one honking great idea -- let's do more of those!)  My thoughts on the suggestions:
* `vla/first`: Makes sense, but doesn't quite generalize to magpis (which includes a bunch of surveys, some from the VLA and some not).
* `nrao/first`: I'm strongly opposed to this one because this is not an NRAO service.
* `image_cutouts/first`: This is correct but perhaps still leaves the namespace too flat?  But I think this is definitely the right place for magpis.
* `radio/first`: Also correct, but like `vla/first` does not comfortably work for magpis.

I assume the reference to a pseudo structure would mean services could appear in multiple locations.  In that case, I think the `first` module could appear as `vla/first`, `radio/first`, and `image_cutouts/first` (as could magpis) and could be actually located in any of those locations.  I guess there would be a mechanism for tagging the services to indicate which pseudo-collections should include them.

rlwastro commented:
I can see the argument for making the namespace structure more hierarchical.  (To quote Tim Peters: Namespaces are one honking great idea -- let's do more of those!)  My thoughts on the suggestions:
* `vla/first`: Makes sense, but doesn't quite generalize to magpis (which includes a bunch of surveys, some from the VLA and some not).
* `nrao/first`: I'm strongly opposed to this one because this is not an NRAO service.
* `image_cutouts/first`: This is correct but perhaps still leaves the namespace too flat?  But I think this is definitely the right place for magpis.
* `radio/first`: Also correct, but like `vla/first` does not comfortably work for magpis.

I assume the reference to a pseudo structure would mean services could appear in multiple locations.  In that case, I think the `first` module could appear as `vla/first`, `radio/first`, and `image_cutouts/first` (as could magpis) and could be actually located in any of those locations.  I guess there would be a mechanism for tagging the services to indicate which pseudo-collections should include them.

ceb8 commented:
My concern with `radio/first` and `image_cutouts/first` would be user confusion since, e.g. alma is radio data but is under alma. But I like the pseudo namespace idea, so that you *could* get to all radio data under radio. In that case it might make sense to have the first location for this code under vla or nrao (whichever @rlwastro thinks is more logical) to keep more consistency throughout the repo.

rlwastro commented:
I'd be fine with vla/first.  As I mentioned above, I am very opposed to nrao/first since this is not an NRAO service.  (I'd be happy to give @ceb8 some background on this opinion in private. 😉)

bsipocz commented:
> I assume the reference to a pseudo structure would mean services could appear in multiple locations. In that case, I think the first module could appear as vla/first, radio/first, and image_cutouts/first (as could magpis) and could be actually located in any of those locations. I guess there would be a mechanism for tagging the services to indicate which pseudo-collections should include them.

> My concern with radio/first and image_cutouts/first would be user confusion since, e.g. alma is radio data but is under alma. But I like the pseudo namespace idea, so that you could get to all radio data under radio. In that case it might make sense to have the first location for this code under vla or nrao (whichever @rlwastro thinks is more logical) to keep more consistency throughout the repo.

yes, the idea with the pseudo namespaces is that all relevant modules will eventually be accessible from there. Discussion on the topic is scattered around in different issues and PRs, the gist of it is the following:

Issues with the fully flat structure:
- the namespace is getting very crowded, which makes functionality discoverability difficult, etc.
- it raises issues with archives providing access to the same data (e.g. mast and esa/hubble, hubble just could not be placed on the top level). 
- modules providing totally different data products are mixed together, we have PI survey, huge data archives, catalogue data access, image services, molecular databases, minor planet orbital elements, etc. 

Due to backward compatibility issues we cannot/should not move the current modules into structure without keeping their access in the main namespace as well, at least for a long deprecation period. This should be the fate of `magpis`.  But for new services that would fall into a structure, we should start to add them into the structured layers from the beginning without waiting for the big refactoring PR to be finished. 

`magpis` is a clear case, so if you all prefer and agree on a `vla` location, and maybe also a `radio` pseudo one, I can open a quick follow-up PR that would bring `magpis` over there, too, so `first` doesn't have to be the only one while waiting for the refactor.

keflavich commented:
We need `image_cutouts/first` and `image_cutouts/magpis`.  `magpis` won't have a second home.  I prefer `image_cutouts` as the home of the `first` code too because we've apparently all agreed on it at some level, and we can therefore get this PR merged and deal with the other namespace links later.

On those namespace links:
For FIRST, how about `radio/vla/first`?    We'll then later have `radio/alma` and `radio/nrao` also.

bsipocz commented:
@keflavich - I like the plan.

However I don't think we need 3 layers for radio, but also I don't mind it if we have 3.

rlwastro commented:
@keflavich Do I need to make the changes to get the module in the right location?  I'm unclear on my responsibility here.  I can put it into `image_cutouts/first` if that's the approved location.

keflavich commented:
Yes, could you move it into `image_cutouts/`?  I think we're ready to merge after that.   Sorry this got delayed by our side discussion

rlwastro commented:
@keflavich I made the necessary changes, and everything is working in my local tests.  But there are errors in the Travis CI run that looks like this:

```
astroquery/image_cutouts/first/core.py:6: in <module>
    from ...query import BaseQuery
E   ValueError: attempted relative import beyond top-level package
```

Here's the link to the test: https://travis-ci.org/github/astropy/astroquery/builds/699092227

Can you tell what the problem is?  I think I have to add the extra dot to get up to the main level of the package.  Is there another way to do this?

rlwastro commented:
I guess I could put an __init__.py file into image_cutouts and import all the base astroquery things that are needed, and then image_cutouts/first could import from .. .  That feels a little hacky, but maybe there is some good reason why Travis CI doesn't like these ... imports?

keflavich commented:
You might need an `__init__.py`, but I don't think importing modules from there will change anything about how you import them within the `first` module.  I probably face the same issues in #1451.

rlwastro commented:
I figure the `image_cutouts/__init__.py` can import all the modules needed from the astroquery directory, and then I can get them from there.  E.g. in `image_cutouts/__init__.py`:
```
from .. import query
```
And then in `image_cutouts/first/core.py`:
```
from .. import query
```
So no import goes up more than one level.  It seems less clear than the direct import from `...` but maybe it is ok.

keflavich commented:
I... don't know if that works?  I'm not sure it should, since it assumes that `image_cutouts` has a namespace defined before `image_cutouts/first`, right?   I don't know how python treats that.

`from ... import query` should work, though.  Does it work when you try it locally?

rlwastro commented:
Yes, everything works fine with the current code.  This is an issue with the Travis CI tests I believe.  I think every `from ... import` with more than two dots is failing in Travis (even though it works fine in reality).

keflavich commented:
I restarted the failing test because I don't believe it.

keflavich commented:
OK, so I can reproduce the errors locally doing `python setup.py test -P image_cutouts/first`.  I'll dig more.

rlwastro commented:
Hmm, here's another thing I can try.  Maybe the imports that cause trouble look like this:
```
from ...query import BaseQuery
```
I can try changing that to 
```
from ... import query
```
and then use `query.BaseQuery`.

keflavich commented:
Solution: just add a *blank* `__init__.py` in `image_cutouts/`

rlwastro commented:
OK!  That's simple if it works.  🤞

rlwastro commented:
Yep, the blank `image_cutouts/__init__.py` file worked like a charm.

bsipocz commented on README.rst:
```diff
@@ -107,6 +107,7 @@ The following modules have been completed using a common API:
   * `CADC <https://astroquery.readthedocs.io/en/latest/cadc/cadc.html>`_: Canadian Astronomy Data Centre.
   * `ESASky <http://astroquery.readthedocs.io/en/latest/esasky/esasky.html>`_: ESASky is a science driven discovery portal providing easy visualizations and full access to the entire sky as observed with ESA Space astronomy missions.
   * `ESO Archive <http://astroquery.readthedocs.io/en/latest/eso/eso.html>`_
+  * `FIRST <http://astroquery.readthedocs.io/en/latest/first/first.html>`_: Faint Images of the Radio Sky at Twenty-cm. 20-cm radio images of the extragalactic sky from the VLA.
```
This link won't exist without adding the narrative docs

rlwastro replied:
Oh thanks, I missed that part.  I assume we need documentation for both the image_cutouts and image_cutouts.first modules.  I will model this after the solarsystem module, which has some nested modules.

A question (for @keflavich too): I see that the solarsystem module has entries at the main level too.  E.g. astroquery.jplhorizons is identical to astroquery.solarsystem.jpl.horizons.  The solarsystem.jpl.horizons `__init__.py` file simply imports everything from the top level astroquery.jplhorizons module.  Should I be doing that for first as well?

keflavich replied:
No, just do the `astroquery.image_cutouts.first`

rlwastro replied:
OK, docs are done and passed the tests.

bsipocz commented:
Thank you @rlwastro!
